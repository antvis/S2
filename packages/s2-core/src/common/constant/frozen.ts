export enum FrozenCellType {
  ROW = 'row',
  COL = 'col',
  TRAILING_ROW = 'trailingRow',
  TRAILING_COL = 'trailingCol',
  SCROLL = 'scroll',
  TOP = 'top',
  BOTTOM = 'bottom',
}

export const FrozenCellGroupMap = {
  [FrozenCellType.ROW]: 'frozenRowGroup',
  [FrozenCellType.COL]: 'frozenColGroup',
  [FrozenCellType.TRAILING_COL]: 'frozenTrailingColGroup',
  [FrozenCellType.TRAILING_ROW]: 'frozenTrailingRowGroup',
  [FrozenCellType.SCROLL]: 'panelScrollGroup',
  [FrozenCellType.TOP]: 'frozenTopGroup',
  [FrozenCellType.BOTTOM]: 'frozenBottomGroup',
};

export interface FrozenOpts {
  frozenRowCount: number;
  frozenColCount: number;
  frozenTrailingRowCount: number;
  frozenTrailingColCount: number;
}

export interface FrozenCellIndex {
  x: number;
  y: number;
}
