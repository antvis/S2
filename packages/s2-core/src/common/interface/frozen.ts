import type { FrozenGroup } from '../../group/frozen-group';
import type { FrozenGroupPosition, FrozenGroupType } from '../constant/frozen';

export type FrozenGroupPositions = Record<
  FrozenGroupPosition,
  {
    width?: number;
    height?: number;
    x?: number;
    y?: number;
    range: number[];
  }
>;

export type FrozenGroups = Record<FrozenGroupType, FrozenGroup>;
