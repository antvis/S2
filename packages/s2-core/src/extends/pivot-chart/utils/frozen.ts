import { Node, type FrozenFacet } from '@antv/s2';

function getNodesByRange(
  nodes: Node[],
  key: 'rowIndex' | 'colIndex',
  minIndex: number,
  maxIndex: number,
) {
  return nodes.filter((node) => node[key] >= minIndex && node[key] <= maxIndex);
}

export function getExtraFrozenColAxisNodes(facet: FrozenFacet, nodes: Node[]) {
  const extraNodes: Node[] = [];

  const { colCount, trailingColCount } = facet.getFrozenOptions();

  if (colCount) {
    const frozenLeafNodes = getNodesByRange(
      nodes,
      'colIndex',
      0,
      colCount - 1,
    )!;

    frozenLeafNodes.forEach((leafNode) => {
      const newLeafNode = leafNode.clone();

      newLeafNode.isFrozenHead = true;
      extraNodes.push(newLeafNode);
    });
  }

  if (trailingColCount) {
    const total = nodes.length;
    const frozenLeafNodes = getNodesByRange(
      nodes,
      'colIndex',
      total - trailingColCount,
      total - 1,
    )!;

    frozenLeafNodes.forEach((leafNode) => {
      const newLeafNode = leafNode.clone();

      newLeafNode.isFrozenTrailing = true;
      extraNodes.push(newLeafNode);
    });
  }

  return extraNodes;
}

export function getExtraFrozenRowAxisNodes(facet: FrozenFacet, nodes: Node[]) {
  const extraNodes: Node[] = [];

  const { start, end } = facet.getCellRange();
  const { rowCount, trailingRowCount } = facet.getFrozenOptions();

  if (rowCount) {
    const frozenLeafNodes = getNodesByRange(
      nodes,
      'rowIndex',
      start,
      start + rowCount - 1,
    )!;

    frozenLeafNodes.forEach((leafNode) => {
      const newLeafNode = leafNode.clone();

      newLeafNode.isFrozenHead = true;
      extraNodes.push(newLeafNode);
    });
  }

  if (trailingRowCount) {
    const frozenLeafNodes = getNodesByRange(
      nodes,
      'rowIndex',
      end - trailingRowCount + 1,
      end,
    )!;

    frozenLeafNodes.forEach((leafNode) => {
      const newLeafNode = leafNode.clone();

      newLeafNode.isFrozenTrailing = true;
      extraNodes.push(newLeafNode);
    });
  }

  return extraNodes;
}
