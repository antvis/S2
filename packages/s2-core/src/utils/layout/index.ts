export { generateId } from './generate-id';
export * from './transform-custom-tree-items';
export * from './frozen';
