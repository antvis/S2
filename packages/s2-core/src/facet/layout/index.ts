import * as _ from 'lodash';
import { SpreadsheetFacet } from '..';
import { BaseDataSet, Pivot } from '../../data-set';
import { EXTRA_FIELD, VALUE_FIELD } from '../../common/constant';
import {
  DrillDownDataCache,
  LayoutResult,
  SpreadsheetFacetCfg,
  StrategyValue,
  ViewMeta,
} from '../../common/interface';
import { Hierarchy } from './hierarchy';
import { getDimsConditionByNode, processCols, processRows } from './util';
import processColNodesCoordinate from './util/process-col-nodes-coordinate';
import { ColsResult } from './util/process-cols';
import processDefaultColWidthByType from './util/process-default-col-width-by-type';
import processRowNodesCoordinate from './util/process-row-nodes-coordinate';
import { RowsResult } from './util/process-rows';
import { BaseParams } from '../../data-set/base-data-set';

/**
 * SpreadSheet's Layout class
 */
export class Layout {
  public facet: SpreadsheetFacet;

  public cfg: SpreadsheetFacetCfg;

  public dataSet: BaseDataSet<BaseParams>;

  public pivot: Pivot;

  public hierarchyType: string;

  public cols: string[];

  public rows: string[];

  public values: string[] | StrategyValue;

  constructor(facet: SpreadsheetFacet) {
    this.facet = facet;
    this.dataSet = facet.getDataset();
    this.pivot = this.dataSet.getPivot();
    this.cfg = facet.cfg;
    this.cols = this.cfg.cols;
    this.rows = this.cfg.rows;
    this.values = this.cfg.values;
    this.hierarchyType = this.cfg.hierarchyType;
  }

  public doLayout(): LayoutResult {
    // 1„ÄÅlayout all nodes in rowHeader and colHeader
    const { rowLeafNodes, rowsHierarchy } = this.getRows();
    const { colLeafNodes, colsHierarchy } = this.getCols(rowsHierarchy);

    // 2„ÄÅhandle col width first by type(spreadsheet or list-sheet | (tree or grid))
    processDefaultColWidthByType(this.facet, this.pivot, colsHierarchy);
    // 3„ÄÅcalculate all nodes coordinate
    processRowNodesCoordinate(
      this.cfg,
      this.facet,
      rowsHierarchy,
      rowLeafNodes,
    );
    processColNodesCoordinate(
      colLeafNodes,
      rowsHierarchy,
      colsHierarchy,
      this.cfg,
      this.facet,
    );

    // Â§ÑÁêÜË°åÂ§¥‰∏≠ÁöÑÂè∂Â≠êËäÇÁÇπÔºåËÆ°ÁÆóÂá∫È´òÂ∫¶‰∏ç‰∏∫0ÁöÑÂè∂Â≠êÁöÑËäÇÁÇπÁöÑrowIndex
    let index = 0;
    rowLeafNodes.forEach((value) => {
      if (value.height !== 0) {
        // eslint-disable-next-line no-param-reassign
        value.rowIndexHeightExist = index;
        index++;
      }
    });
    const dataSet = this.dataSet;
    const ss = this.facet.spreadsheet;

    function getViewMeta(rowIndex: number, colIndex: number): ViewMeta {
      const i = rowIndex || 0;
      const j = colIndex || 0;
      const row = rowLeafNodes[i];
      const col = colLeafNodes[j];
      if (!row || !col) {
        return null;
      }
      // È´òÂ∫¶Á≠â‰∫é0 Áõ¥Êé•ËøîÂõûÁ©∫Ôºå‰∏çÂàõÂª∫‰ªª‰ΩïÁöÑcellÂíåÊï∞ÊçÆËÆ°ÁÆó
      if (row.isHide()) {
        return null;
      }
      // Ê†ëÁä∂Ë°®Ê†ºÊ≤°ÊúâÂ∞èËÆ°Ë°åÔºåÈúÄË¶ÅÂà§Êñ≠ËØ•ÂçïÂÖÉÊ†ºÊòØÂê¶Â±ïÁ§∫ËÅöÂêàÊï∞ÊçÆ
      const rowTotalsConfig = dataSet.pivot.getTotalsConfig(row.key);
      const colTotalsConfig = dataSet.pivot.getTotalsConfig(col.key);
      // ÊòØÂê¶ÊòØÊ†ëÁä∂Â∏ÉÂ±Ä
      const isInTree = ss.isHierarchyTreeType();
      // ÊòØÂê¶ÊòØÁà∂ËäÇÁÇπÂØπÂ∫îÁöÑÂçïÂÖÉÊ†º
      // Êî∂Ëµ∑Êó∂ÔºåÂà§Êñ≠ÊòØÂê¶ÊòØÂè∂Â≠êËäÇÁÇπÔºõÂ±ïÂºÄÊó∂Âà§Êñ≠ÊòØÂê¶Êúâ
      const isParentInTree = !row.isLeaf;
      // Áà∂ËäÇÁÇπÊòØÂê¶ÈúÄË¶ÅÂ±ïÁ§∫Â∞èËÆ° -- tree mode don't need subTotals
      const isSubTotalsInTree =
        isInTree && rowTotalsConfig.showSubTotals && isParentInTree;
      // check if the node is totals(grandTotal or subTotal)
      const isTotals: boolean =
        isSubTotalsInTree || row.isTotals || col.isTotals;
      // grand totals node
      const isGrandTotals = row.isGrandTotals || col.isGrandTotals;
      const isSubTotals = row.isSubTotals || col.isSubTotals;
      const isValueInColsTotal = !ss.options.valueInCols && col.isGrandTotals;
      // isSubTotalsInTree use to get the whole path of node by force
      const rowQuery = getDimsConditionByNode(row, isSubTotalsInTree);
      // Êï∞ÂÄºÊåÇË°åÂ§¥‰∏îÊúâÂàóÊÄªËÆ°ÁöÑÁâπÊÆäÊÉÖÂÜµÔºåÈúÄË¶ÅÂº∫Âà∂ËøîÂõûnodeË∑ØÂæÑ
      const colQuery = getDimsConditionByNode(col, isValueInColsTotal);
      const dataQuery = isValueInColsTotal
        ? _.merge({}, rowQuery)
        : _.merge({}, rowQuery, colQuery);

      let data = [];
      if (isInTree && isParentInTree) {
        // tree mode & collapse
        // Áà∂ËäÇÁÇπ
        if (rowTotalsConfig.showSubTotals) {
          data = dataSet.getData(dataQuery, {
            row: {
              isTotals: true,
              isGrandTotals: row.isGrandTotals,
              isSubTotals: true,
            },
            col: {
              isTotals: col.isTotals,
              isGrandTotals: col.isGrandTotals,
              isSubTotals: col.isSubTotals,
            },
          });
        } else {
          // ÊòØüå≤Áä∂‰∏î‰∏çÈúÄË¶ÅÂ±ïÁ§∫Â∞èËÆ°Ë°åÔºåÈúÄË¶ÅÊúâ‰∏§‰∏™Â§ÑÁêÜÂú∫ÊôØ
          // 1. ËØ•row id‰∏ãÊúâ‰∏ãÈíªÁª¥Â∫¶ÔºåÂàôËøôË°åÊòæÁ§∫ÂéüÂßãÁöÑÂÄº(‰∏çÊåâÂ∞èËÆ°Â§ÑÁêÜ)
          // 2. Ê≤°Êúâ‰∏ãÈíªÁª¥Â∫¶ÔºåÁî±‰∫éÊ≤°ÊúâÈÖçÂ∞èËÆ° ÊòæÁ§∫‰∏∫ -
          const drillDownDataCache = ss.store.get(
            'drillDownDataCache',
            [],
          ) as DrillDownDataCache[];
          const cache = drillDownDataCache.find((dc) => dc.rowId === row.id);
          if (cache) {
            data = dataSet.getData(dataQuery);
          } else {
            data = [];
          }
        }
      } else {
        data = dataSet.getData(dataQuery, {
          row: {
            isTotals: row.isTotals,
            isGrandTotals: row.isGrandTotals,
            isSubTotals: row.isSubTotals,
            otherQuery: colQuery,
            isCollapsedHasTotals:
              rowTotalsConfig.showSubTotals && row.isCollapsed,
          },
          col: {
            isTotals: col.isTotals,
            isGrandTotals: col.isGrandTotals,
            isSubTotals: col.isSubTotals,
            otherQuery: rowQuery,
            isCollapsedHasTotals:
              colTotalsConfig.showSubTotals && col.isCollapsed,
          },
        });
      }

      // mark grand totals node in origin data obj
      _.each(data, (d) => {
        // eslint-disable-next-line no-param-reassign
        d.isGrandTotals = isGrandTotals;
        // eslint-disable-next-line no-param-reassign
        d.isSubTotals = isSubTotals || isSubTotalsInTree;
      });
      let valueField = '';
      let fieldValue = null;
      const realData = _.get(data, [0], {});
      if (!_.isEmpty(realData)) {
        if (_.has(realData, EXTRA_FIELD) || _.has(realData, VALUE_FIELD)) {
          valueField = _.get(realData, [EXTRA_FIELD], '');
          fieldValue = _.get(realData, [VALUE_FIELD], null);
        }
      } else {
        // Êï∞ÊçÆÊü•ËØ¢‰∏∫Á©∫ÔºåÈúÄË¶ÅÈªòËÆ§Â∏¶‰∏äÂèØËÉΩÂ≠òÂú®ÁöÑvalueField
        valueField = _.get(dataQuery, [EXTRA_FIELD], '');
      }
      return {
        spreadsheet: ss,
        x: col.x,
        y: row.y,
        width: col.width,
        height: row.height,
        data,
        rowIndex: i,
        colIndex: j,
        isTotals,
        valueField,
        fieldValue,
        rowQuery,
        colQuery,
        rowId: row.id,
        colId: col.id,
        rowIndexHeightExist: row.rowIndexHeightExist,
      } as ViewMeta;
    }

    const layoutResult = {
      colNodes: colsHierarchy.getNodes(),
      colsHierarchy,
      rowNodes: rowsHierarchy.getNodes(),
      rowsHierarchy,
      rowLeafNodes,
      colLeafNodes,
      getViewMeta,
      spreadsheet: ss,
    } as LayoutResult;
    const callback = this.cfg?.layoutResult;
    return callback ? callback(layoutResult) : layoutResult;
  }

  protected getCols(rowsHierarchy: Hierarchy): ColsResult {
    return processCols(this.pivot, this.cfg, this.cols);
  }

  protected getRows(): RowsResult {
    return processRows(this.pivot, this.cfg, this.rows, this.facet);
  }
}
