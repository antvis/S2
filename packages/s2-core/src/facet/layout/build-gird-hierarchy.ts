import { isEmpty, isUndefined, uniqWith } from 'lodash';
import { FieldValue, GridHeaderParams } from '@/facet/layout/interface';
import { TotalMeasure } from '@/facet/layout/total-measure';
import { layoutArrange } from '@/facet/layout/layout-hooks';
import { getDimsCondition } from '@/utils/layout/get-dims-condition-by-node';
import { addTotals } from '@/utils/layout/add-totals';
import { generateHeaderNodes } from '@/utils/layout/generate-header-nodes';
import { EXTRA_FIELD } from '@/common/constant';
import { SpreadSheetFacetCfg } from '@/common/interface';

const hideMeasureColumn = (
  fieldValues: FieldValue[],
  field: string,
  cfg: SpreadSheetFacetCfg,
) => {
  const hideMeasure = cfg.colCfg?.hideMeasureColumn ?? false;
  const valueInCol = cfg.dataSet.fields.valueInCols;
  for (const value of fieldValues) {
    if (hideMeasure && valueInCol && field === EXTRA_FIELD) {
      fieldValues.splice(fieldValues.indexOf(value), 1);
    }
  }
};

/**
 * Build grid hierarchy in rows or columns
 *
 * @param params
 */
export const buildGridHierarchy = (params: GridHeaderParams) => {
  const {
    addTotalMeasureInTotal,
    addMeasureInTotalQuery,
    parentNode,
    currentField,
    fields,
    facetCfg,
    hierarchy,
  } = params;

  const index = fields.indexOf(currentField);

  const { dataSet, values, spreadsheet } = facetCfg;
  const fieldValues: FieldValue[] = [];

  let query = {};
  if (parentNode.isTotals) {
    // add total measures
    if (addTotalMeasureInTotal) {
      query = getDimsCondition(parentNode.parent, true);
      // add total measures
      fieldValues.push(...values.map((v) => new TotalMeasure(v)));
    }
  } else {
    // field(dimension)'s all values
    query = getDimsCondition(parentNode, true);

    const dimValues = dataSet.getDimensionValues(currentField, query);

    const arrangedValues = layoutArrange(
      dimValues,
      facetCfg,
      parentNode,
      currentField,
    );
    fieldValues.push(...(arrangedValues || []));

    // add skeleton for empty data

    const fieldName = dataSet.getFieldName(currentField);

    if (isEmpty(fieldValues)) {
      if (currentField === EXTRA_FIELD) {
        fieldValues.push(...dataSet.fields?.values);
      } else {
        fieldValues.push(fieldName);
      }
    }
    // hide measure in columns
    hideMeasureColumn(fieldValues, currentField, facetCfg);
    // add totals if needed
    addTotals({
      currentField,
      lastField: fields[index - 1],
      isFirstField: index === 0,
      fieldValues,
      spreadsheet,
    });
  }

  const hiddenColumnsDetail = spreadsheet.store.get('hiddenColumnsDetail');
  const isEqualValueLeafNode =
    uniqWith(spreadsheet.getColumnLeafNodes(), (prev, next) => {
      return prev.value === next.value;
    }).length === 1;

  const displayFieldValues = fieldValues.filter((value) => {
    // 去除多余的节点
    if (isUndefined(value)) {
      return false;
    }
    if (isEmpty(hiddenColumnsDetail)) {
      return true;
    }
    return hiddenColumnsDetail.every((detail) => {
      return detail.hideColumnNodes.every((node) => {
        // 有数值字段 (hideMeasureColumn: true) 隐藏父节点
        // 多列头场景(数值挂列头, 为隐藏数值列头, 自定义目录多指标等) 叶子节点是数值, 叶子节点的文本内容都一样, 需要额外比较父级节点的id是否相同, 确定到底渲染哪一列
        const isMeasureField = node.field === EXTRA_FIELD;
        if (isMeasureField || isEqualValueLeafNode) {
          return (
            node.parent.id !== parentNode.id && node.parent.value !== value
          );
        }
        // 没有数值字段 (hideMeasureColumn: false) 隐藏自己即可
        return node.value !== value;
      });
    });
  });

  generateHeaderNodes({
    currentField,
    fields,
    fieldValues: displayFieldValues,
    facetCfg,
    hierarchy,
    parentNode,
    level: index,
    query,
    addMeasureInTotalQuery,
    addTotalMeasureInTotal,
  });
};
