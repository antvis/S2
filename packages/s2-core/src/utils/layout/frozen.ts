import { S2TableSheetOptions } from 'src/common/interface';

export const getValidFrozenOptions = (
  opts: S2TableSheetOptions,
  colLength: number,
  dataLength = 0,
) => {
  // 如果没有传行列冻结选项，提前返回
  if (!Object.values(opts).find((item) => item > 0)) {
    return opts;
  }

  const newOpts = {
    ...opts,
  };

  if (newOpts.frozenColCount >= colLength) {
    newOpts.frozenColCount = colLength;
  }

  const remainFrozenColCount = colLength - newOpts.frozenColCount;

  if (newOpts.frozenTrailingColCount > remainFrozenColCount) {
    newOpts.frozenTrailingColCount = remainFrozenColCount;
  }

  if (newOpts.frozenRowCount >= dataLength) {
    newOpts.frozenRowCount = dataLength;
  }

  const remainFrozenRowCount = dataLength - newOpts.frozenRowCount;
  if (newOpts.frozenTrailingRowCount > remainFrozenRowCount) {
    newOpts.frozenTrailingRowCount = remainFrozenRowCount;
  }

  return newOpts;
};
