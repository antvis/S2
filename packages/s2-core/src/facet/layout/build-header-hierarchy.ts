import { BuildHeaderParams, BuildHeaderResult } from '@/facet/layout/interface';
import { Hierarchy } from '@/facet/layout/hierarchy';
import { Node } from '@/facet/layout/node';
import { buildRowTreeHierarchy } from '@/facet/layout/build-row-tree-hierarchy';
import { buildGridHierarchy } from '@/facet/layout/build-gird-hierarchy';
import { buildTableHierarchy } from '@/facet/layout/build-table-hierarchy';
import { PivotDataSet } from '@/data-set';
import { buildRowCustomTreeHierarchy } from '@/facet/layout/build-row-custom-tree-hierarchy';
import { SpreadSheet } from '@/sheet-type';
import { SpreadSheetFacetCfg } from '@/common/interface';

interface HeaderParams {
  isValueInCols: boolean;
  isPivotMode: boolean;
  moreThanOneValue: boolean;
  hierarchy: Hierarchy;
  rootNode: Node;
  spreadsheet: SpreadSheet;
  facetCfg: SpreadSheetFacetCfg;
  fields: string[];
  isRowHeader: boolean;
}

const handleGridRowColHierarchy = (params: HeaderParams) => {
  const {
    isValueInCols,
    moreThanOneValue,
    rootNode,
    facetCfg,
    hierarchy,
    fields,
    isRowHeader,
  } = params;
  // add new total measure in total node
  let addTotalMeasureInTotal: boolean;
  // add measure info in total query
  let addMeasureInTotalQuery: boolean;
  if (isRowHeader) {
    addTotalMeasureInTotal = !isValueInCols && moreThanOneValue;
    addMeasureInTotalQuery = !isValueInCols && !moreThanOneValue;
  } else {
    addTotalMeasureInTotal = isValueInCols && moreThanOneValue;
    addMeasureInTotalQuery = isValueInCols && !moreThanOneValue;
  }

  buildGridHierarchy({
    addTotalMeasureInTotal,
    addMeasureInTotalQuery,
    parentNode: rootNode,
    currentField: fields[0],
    fields,
    facetCfg,
    hierarchy,
  });
};

const handleCustomTreeRowHierarchy = (params: HeaderParams) => {
  const { facetCfg, rootNode, hierarchy } = params;
  const customTreeItems = facetCfg.dataSet.fields.customTreeItems;
  // row custom tree header
  buildRowCustomTreeHierarchy({
    customTreeItems: customTreeItems,
    facetCfg,
    level: 0,
    parentNode: rootNode,
    hierarchy,
  });
};

const handleTreeRowHierarchy = (params: HeaderParams) => {
  const { facetCfg, rootNode, hierarchy } = params;
  const { hierarchyType, rows, dataSet } = facetCfg;
  if (hierarchyType === 'tree') {
    // row tree hierarchy(value must stay in colHeader)
    buildRowTreeHierarchy({
      level: 0,
      currentField: rows[0],
      pivotMeta: (dataSet as PivotDataSet).rowPivotMeta,
      facetCfg,
      parentNode: rootNode,
      hierarchy,
    });
  } else {
    handleCustomTreeRowHierarchy(params);
  }
};

const handleRowHeaderHierarchy = (params: HeaderParams) => {
  // 只有透视表有行头
  const { spreadsheet } = params;
  if (spreadsheet.isHierarchyTreeType()) {
    handleTreeRowHierarchy(params);
  } else {
    handleGridRowColHierarchy(params);
  }
};

const handleColHeaderHierarchy = (params: HeaderParams) => {
  const { isPivotMode, hierarchy, rootNode, facetCfg } = params;
  if (isPivotMode) {
    handleGridRowColHierarchy(params);
  } else {
    buildTableHierarchy({
      parentNode: rootNode,
      hierarchy,
      facetCfg,
    });
  }
};

/**
 * Header Hierarchy
 * - row header
 *   - tree layout
 *     - custom tree layout
 *   - grid layout
 *   - table layout
 * - col header
 *   - grid layout
 *     - single value
 *       - total + sub_total
 *     - more than one value
 *       - total + sub_total
 *         - separate by values
 *   - table layout
 * @param params
 */
export const buildHeaderHierarchy = (
  params: BuildHeaderParams,
): BuildHeaderResult => {
  const { isRowHeader, facetCfg } = params;
  const { spreadsheet, rows = [], columns = [] } = facetCfg;
  const isValueInCols = spreadsheet.dataCfg.fields.valueInCols;
  const isPivotMode = spreadsheet.isPivotMode();
  const moreThanOneValue = facetCfg.dataSet.moreThanOneValue();
  const rootNode = Node.rootNode();
  const hierarchy = new Hierarchy();
  const headParams = {
    isValueInCols,
    isPivotMode,
    moreThanOneValue,
    rootNode,
    hierarchy,
    spreadsheet,
    facetCfg,
    fields: isRowHeader ? rows : columns,
    isRowHeader,
  } as HeaderParams;
  if (isRowHeader) {
    handleRowHeaderHierarchy(headParams);
  } else {
    handleColHeaderHierarchy(headParams);
  }

  const getLeafNodes = () => {
    if (!isRowHeader) return hierarchy.getLeaves();
    return spreadsheet.isHierarchyTreeType()
      ? hierarchy.getNodes()
      : hierarchy.getLeaves();
  };

  return {
    hierarchy,
    leafNodes: getLeafNodes(),
  } as BuildHeaderResult;
};
