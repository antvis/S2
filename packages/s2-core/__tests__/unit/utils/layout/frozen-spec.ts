import { getValidFrozenOptions } from '../../../../src/utils/layout/frozen';

describe('frozen test', () => {
  test('should get valid frozen options', () => {
    expect(
      getValidFrozenOptions(
        {
          colCount: 0,
          rowCount: 0,
          trailingColCount: 0,
          trailingRowCount: 0,
        },
        10,
        10,
      ),
    ).toMatchSnapshot();

    expect(
      getValidFrozenOptions(
        {
          colCount: 1,
          rowCount: 2,
          trailingColCount: 3,
          trailingRowCount: 4,
        },
        10,
        10,
      ),
    ).toMatchSnapshot();

    expect(
      getValidFrozenOptions(
        {
          colCount: 11,
          rowCount: 11,
          trailingColCount: 11,
          trailingRowCount: 11,
        },
        10,
        10,
      ),
    ).toMatchSnapshot();
  });
});
