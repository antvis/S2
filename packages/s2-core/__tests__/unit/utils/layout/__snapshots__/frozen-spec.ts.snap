// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`frozen test should get valid frozen options 1`] = `
Object {
  "colCount": 0,
  "rowCount": 0,
  "trailingColCount": 0,
  "trailingRowCount": 0,
}
`;

exports[`frozen test should get valid frozen options 2`] = `
Object {
  "colCount": 1,
  "rowCount": 2,
  "trailingColCount": 3,
  "trailingRowCount": 4,
}
`;

exports[`frozen test should get valid frozen options 3`] = `
Object {
  "colCount": 10,
  "rowCount": 10,
  "trailingColCount": 0,
  "trailingRowCount": 0,
}
`;
